flepimop/gempyor_pkg/src/gempyor/sync/_sync.py

```diff
@@ -1,336 +1,528 @@
-        return [
-            "--exclude" if filter[0] == "-" else "--include",
-            '"{}"'.format(filter[1]),
-        ]
+    def _check_at_least_one_bucket(self) -> "S3SyncModel":
+        """
+        Check that one of the source or target is an S3 bucket.
 
-    @staticmethod
-    def _dryrun(dry: bool) -> list[str]:
-        return ["--dryrun"] if dry else []
+        Raises:
+            ValueError: If neither source nor target is an S3 bucket.
+        """
+        if self.target.startswith("s3://") or self.source.startswith("s3://"):
+            return self
+        raise ValueError(
+            "At least one of `source` or `target` must be "
+            "an s3 bucket, as indicated by a `s3://` prefix"
+        )
 
     @staticmethod
-    def _cmd() -> list[str]:
-        return ["aws", "s3", "sync"]
+    def _formatter(f: FilterParts) -> list[str]:
+        if f[0] == "s":
+            return ["--exclude", '"*"', "--include", f'"*{f[1]}*"']
+        return ["--exclude" if f[0] == "-" else "--include", f'"{f[1]}"']
 
     def _sync_pydantic(
-        self, sync_options: SyncOptions = SyncOptions(), verbosity: int = 0
+        self, sync_options: SyncOptions, verbosity: int = 0
     ) -> CompletedProcess:
-        inner_paths = [
-            str(_override_or_val(sync_options.source_override, self.source)) + "/",
-            str(_override_or_val(sync_options.target_override, self.target)) + "/",
-        ]
-        match self.s3:
-            case "source":
-                inner_paths[0] = "s3:" + inner_paths[0]
-            case "target":
-                inner_paths[1] = "s3:" + inner_paths[1]
-            case "both":
-                inner_paths = ["s3:" + p for p in inner_paths]
-
+        logger = get_script_logger(__name__, verbosity)
+        inner_paths = list(
+            _resolve_s3_inner_paths(
+                [sync_options.source(self.source), sync_options.target(self.target)]
+            )
+        )
+        logger.debug("Resolved paths: %s", str(inner_paths))
         if sync_options.reverse:
             inner_paths.reverse()
+            logger.debug("Reversed paths, now resolved: %s", str(inner_paths))
         inner_filter = self.format_filters(
-            sync_options.filter_override,
-            sync_options.filter_prefix,
-            sync_options.filter_suffix,
+            overrides=sync_options.filter_override,
+            prefix=sync_options.filter_prefix,
+            suffix=sync_options.filter_suffix,
             reverse=True,
         )
-        testcmd = (
-            self._cmd() + self._dryrun(sync_options.dryrun) + inner_filter + inner_paths
+        logger.debug("Resolved filters: %s", str(inner_filter))
+        cmd = (
+            [
+                _shutil_which("aws"),
+                "s3",
+                (
+                    "cp"
+                    if any(os.path.exists(p) and os.path.isfile(p) for p in inner_paths)
+                    else "sync"
+                ),
```

Comment:
do we actually need to switch between sync vs cp?

Reply:
I do not remember OTOH why I made 7cd6c6521534b9140bfb90bd2a62c5699c2b0784, but I'll double check.